File: apps/opensible-server/services/webhook_dispatcher.py
"""Outbound webhook dispatcher (Fase 1 — UC 95).

Storage: a JSON file under DATA_DIR (``webhooks.json``), consistent with the
other config JSON files. Dispatch is fire-and-forget: each matching webhook is
POSTed in its own thread with HMAC-SHA256 signature and 3 retries.
"""
from __future__ import annotations

import hashlib
import hmac
import json
import logging
import threading
import time
from pathlib import Path
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)

_DATA_DIR: Optional[Path] = None
_LOCK = threading.Lock()


def _store_path() -> Path:
    global _DATA_DIR
    if _DATA_DIR is None:
        try:
            import app as _app
            _DATA_DIR = Path(getattr(_app, "DATA_DIR", "data"))
        except Exception:
            _DATA_DIR = Path("data")
    return _DATA_DIR / "webhooks.json"


def set_data_dir(path: Path) -> None:
    global _DATA_DIR
    _DATA_DIR = Path(path)


def load_webhooks() -> List[Dict[str, Any]]:
    p = _store_path()
    try:
        if p.exists():
            with open(p, "r", encoding="utf-8") as f:
                data = json.load(f)
                if isinstance(data, list):
                    return data
    except Exception as e:
        logger.error(f"[webhooks] failed to load: {e}")
    return []


def _save_webhooks(webhooks: List[Dict[str, Any]]) -> None:
    p = _store_path()
    try:
        p.parent.mkdir(parents=True, exist_ok=True)
        with open(p, "w", encoding="utf-8") as f:
            json.dump(webhooks, f, indent=2)
    except Exception as e:
        logger.error(f"[webhooks] failed to save: {e}")


def create_webhook(url: str, events: List[str], secret: str = "") -> Dict[str, Any]:
    import uuid as _uuid
    wh = {
        "id": str(_uuid.uuid4()),
        "url": url,
        "secret": secret,
        "events": events or [],
        "enabled": True,
        "created_at": time.time(),
    }
    with _LOCK:
        whs = load_webhooks()
        whs.append(wh)
        _save_webhooks(whs)
    return wh


def update_webhook(webhook_id: str, updates: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    with _LOCK:
        whs = load_webhooks()
        for wh in whs:
            if wh.get("id") == webhook_id:
                for k in ("url", "secret", "events", "enabled"):
                    if k in updates:
                        wh[k] = updates[k]
                _save_webhooks(whs)
                return wh
    return None


def delete_webhook(webhook_id: str) -> bool:
    with _LOCK:
        whs = load_webhooks()
        nxt = [w for w in whs if w.get("id") != webhook_id]
        if len(nxt) != len(whs):
            _save_webhooks(nxt)
            return True
    return False


def _sign(secret: str, body: bytes) -> str:
    return hmac.new(secret.encode("utf-8"), body, hashlib.sha256).hexdigest()


def _deliver(wh: Dict[str, Any], event: str, payload: Dict[str, Any]) -> None:
    import requests

    body = json.dumps(payload).encode("utf-8")
    headers = {
        "Content-Type": "application/json",
        "User-Agent": "radas-webhook/1.0",
        "X-Radas-Event": event,
    }
    if wh.get("secret"):
        headers["X-Radas-Signature"] = "sha256=" + _sign(wh["secret"], body)

    last_err = None
    for attempt in range(3):
        try:
            r = requests.post(wh["url"], data=body, headers=headers, timeout=5)
            if 200 <= r.status_code < 300:
                return
            last_err = f"http {r.status_code}"
        except Exception as e:
            last_err = str(e)
        time.sleep(1.5 * (attempt + 1))
    logger.warning(f"[webhooks] delivery failed {event} -> {wh['url']}: {last_err}")


def dispatch_event(event: str, payload: Dict[str, Any]) -> int:
    """Fire ``event`` to every enabled webhook subscribed to it (async)."""
    sent = 0
    for wh in load_webhooks():
        if not wh.get("enabled"):
            continue
        if event not in (wh.get("events") or []):
            continue
        threading.Thread(target=_deliver, args=(wh, event, payload), daemon=True).start()
        sent += 1
    if sent:
        logger.info(f"[webhooks] dispatched {event} to {sent} webhook(s)")
    return sent

File: apps/server/lib/radas_ai/webhook_dispatcher.ex
defmodule RadasAI.WebhookDispatcher do
  @moduledoc """
  Port of `services/webhook_dispatcher.py` — outbound webhooks stored in the
  kv_store `webhooks` scope (list under the "default" key). Delivery is async
  (Task), HMAC-SHA256 signed, 3 attempts, 5s timeout, best-effort.
  """

  import RadasAI.DB

  alias RadasAI.KV

  defp load_webhooks do
    case KV.load("webhooks") do
      v when is_list(v) -> Enum.filter(v, &is_map/1)
      _ -> []
    end
  end

  defp save_webhooks(items), do: KV.save("webhooks", Enum.filter(items, &is_map/1))

  @doc "Create one webhook subscription (Python create_webhook)."
  @spec create_webhook(String.t(), [String.t()], String.t()) :: map()
  def create_webhook(url, events, secret \\ "") do
    now = System.system_time(:second)

    wh = %{
      "id" => Ecto.UUID.generate(),
      "url" => to_string(url),
      "events" => Enum.map(List.wrap(events || []), &to_string/1),
      "secret" => to_string(secret || ""),
      "enabled" => true,
      "created_at" => now,
      "updated_at" => now
    }

    save_webhooks([wh | load_webhooks()])
    wh
  end

  @doc "Update one webhook; returns nil when not found."
  @spec update_webhook(String.t(), map()) :: map() | nil
  def update_webhook(webhook_id, updates) do
    items = load_webhooks()

    {updated, rest} =
      Enum.split_with(items, &(&1["id"] == webhook_id))

    case updated do
      [] ->
        nil

      [wh] ->
        wh =
          wh
          |> Map.merge(Map.new(updates || %{}, fn {k, v} -> {to_string(k), v} end))
          |> Map.put("updated_at", System.system_time(:second))

        save_webhooks([wh | rest])
        wh
    end
  end

  @doc "Delete one webhook; returns false when not found."
  @spec delete_webhook(String.t()) :: boolean()
  def delete_webhook(webhook_id) do
    items = load_webhooks()
    rest = Enum.reject(items, &(&1["id"] == webhook_id))

    if length(rest) == length(items) do
      false
    else
      save_webhooks(rest)
      true
    end
  end

  @doc """
  Fire `event` to every enabled webhook subscribed to it (async, fail-open).
  Returns the number of dispatches started (Python dispatch_event).
  """
  @spec dispatch_event(String.t(), map()) :: integer()
  def dispatch_event(event, payload) do
    body = Jason.encode!(payload || %{})

    targets =
      Enum.filter(load_webhooks(), fn wh ->
        wh["enabled"] and event in List.wrap(wh["events"] || [])
      end)

    Enum.each(targets, fn wh ->
      Task.start(fn -> deliver(wh, event, body) end)
    end)

    length(targets)
  rescue
    _ -> 0
  end

  defp deliver(wh, event, body) do
    headers = %{
      "content-type" => "application/json",
      "user-agent" => "radas-webhook/1.0",
      "x-radas-event" => event
    }

    headers =
      if wh["secret"] not in [nil, ""] do
        Map.put(headers, "x-radas-signature", "sha256=" <> sign(wh["secret"], body))
      else
        headers
      end

    Enum.reduce_while(1..3, :ok, fn attempt, _acc ->
      case Req.post(wh["url"], body: body, headers: headers, retry: false) do
        {:ok, %Req.Response{status: s}} when s in 200..299 ->
          {:halt, :ok}

        _ ->
          if attempt == 3, do: {:halt, :error}, else: {:cont, :retry}
      end
    end)

    :ok
  rescue
    _ -> :error
  end

  defp sign(secret, body) do
    :crypto.mac(:hmac, :sha256, secret, body) |> Base.encode16(case: :lower)
  end

  @doc "Redact a probe detail line (Python _safe_probe_detail lives in byoc; shared helper)."
  @spec redact_detail(String.t()) :: String.t()
  def redact_detail(value) do
    text = to_string(value || "")

    Enum.reduce(["password", "token", "secret", "apikey", "authorization"], text, fn marker, acc ->
      Regex.replace(
        ~r/(#{marker}\s*[=:]\s*)[^,;\s}]+/i,
        acc,
        "\\1[REDACTED]"
      )
    end)
    |> String.slice(0, 200)
  end
end
